d2bs/kolbot/libs/OOG.js

```diff
@@ -1450,15 +1463,77 @@ const Starter = {
 		}
 	},
 
+	receiveCopyData: function (mode, msg) {
+		let obj;
+
+		msg === "Handle" && typeof mode === "number" && (Starter.handle = mode);
+
+		switch (mode) {
+		case 1: // JoinInfo
+			obj = JSON.parse(msg);
+			Object.assign(Starter.joinInfo, obj);
+
+			break;
+		case 2: // Game info
+			print("Recieved Game Info");
+			obj = JSON.parse(msg);
+			Object.assign(Starter.gameInfo, obj);
+
+			break;
+		case 3: // Game request
+			// Don't let others join mule/torch/key/gold drop game
+			if (AutoMule.inGame || Gambling.inGame || TorchSystem.inGame || CraftingSystem.inGame) {
+				break;
+			}
+
+			if (Object.keys(Starter.gameInfo).length) {
+				obj = JSON.parse(msg);
+
+				if ([sdk.game.profiletype.TcpIpHost, sdk.game.profiletype.TcpIpJoin].includes(Profile().type)) {
+					me.gameReady && D2Bot.joinMe(obj.profile, me.gameserverip.toString(), "", "", isUp);
+				} else {
+					if (me.gameReady) {
+						D2Bot.joinMe(obj.profile, me.gamename.toLowerCase(), "", me.gamepassword.toLowerCase(), isUp);
+					} else {
+						D2Bot.joinMe(obj.profile, Starter.gameInfo.gameName.toLowerCase(), Starter.gameCount, Starter.gameInfo.gamePass.toLowerCase(), isUp);
+					}
+				}
+			}
+
+			break;
+		case 4: // Heartbeat ping
+			msg === "pingreq" && sendCopyData(null, me.windowtitle, 4, "pingrep");
+
+			break;
+		case 61732: // Cached info retreival
+			msg !== "null" && (Starter.gameInfo.crashInfo = JSON.parse(msg));
+
+			break;
+		case 1638: // getProfile
+			try {
+				obj = JSON.parse(msg);
+				Starter.profileInfo.profile = me.profile;
+				Starter.profileInfo.account = obj.account;
+				Starter.profileInfo.charName = obj.Character;
+				obj.Realm = obj.Realm.toLowerCase();
+				Starter.profileInfo.realm = ["east", "west"].includes(obj.Realm) ? "us" + obj.Realm : obj.Realm;
+			} catch (e) {
+				print(e);
+			}
+
+			break;
+		}
+	},
+
 	randomString: function (len, useNumbers = false) {
 		len === undefined && (len = rand(5, 14));
 
 		let rval = "",
 			letters = useNumbers ? "abcdefghijklmnopqrstuvwxyz0123456789" : "abcdefghijklmnopqrstuvwxyz";
-
+		let strLen = letters.length;
 
 		for (let i = 0; i < len; i += 1) {
-			rval += letters[rand(0, letters.length - 1)];
+			rval += letters[rand(0, strLen)];
```

Comment:
Careful, rand() is not js random() but according to d2bs doc:

> Uses C [rand()](https://github.com/noah-/d2bs#group__globalFunctions_1ga6cbf7046bc4629083c0d50f6c5ce8843) when out of game, and D2GAME_D2Rand with the game seed while in game.

accessing letters out-of-bounds randomly would be a problem. So either change this back to `length - 1` or perform a check to make sure `length` is excluded.